extensions/stackdriver/edges/edge_reporter.cc

```diff
@@ -0,0 +1,149 @@
+  instance->set_cluster_name(
+      node_info.platform_metadata().at(Common::kGCPClusterNameKey));
+  instance->set_owner_uid(node_info.owner());
+  instance->set_workload_name(node_info.workload_name());
+  instance->set_workload_namespace(node_info.namespace_());
+};
+
+}  // namespace
+
+EdgeReporter::EdgeReporter(const ::wasm::common::NodeInfo& local_node_info,
+                           std::unique_ptr<MeshEdgesServiceClient> edges_client,
+                           TimestampFn now)
+    : now_(now) {
+  current_request_ = std::make_unique<ReportTrafficAssertionsRequest>();
+
+  const auto& project_id =
+      local_node_info.platform_metadata().at(Common::kGCPProjectKey);
+  current_request_->set_parent("projects/" + project_id);
+
+  // TODO(dougreid): figure out how to get the real mesh uid here.
+  // Using: //.../projects/<project_id>/<location>/meshes/<cluster> as a
+  // placeholder.
+  std::string location =
+      local_node_info.platform_metadata().at(Common::kGCPClusterLocationKey);
+  std::string cluster =
+      local_node_info.platform_metadata().at(Common::kGCPClusterNameKey);
+  absl::StrAppend(current_request_->mutable_mesh_uid(),
+                  "//cloudresourcemanager.googleapis.com/projects/", project_id,
+                  "/", location, "/meshes/", cluster);
+
+  instanceFromMetadata(local_node_info, &node_instance_);
+
+  edges_client_ = std::move(edges_client);
+};
+
+// ONLY inbound
+void EdgeReporter::addEdge(const ::Wasm::Common::RequestInfo& request_info,
+                           const ::wasm::common::NodeInfo& peer_node_info) {
+  auto peer = current_peers_.find(peer_node_info.node_key());
+  if (peer != current_peers_.end()) {
+    return;
+  }
+  current_peers_.emplace(peer_node_info.node_key());
+
+  auto* traffic_assertions = current_request_->mutable_traffic_assertions();
+  auto* edge = traffic_assertions->Add();
+
+  // TODO(douglas-reid): use the short name for the destination service when
+  // available Right now, this uses destination host instead.
+  edge->set_destination_service_name(request_info.destination_service_host);
+  edge->set_destination_service_namespace(node_instance_.workload_namespace());
+  instanceFromMetadata(peer_node_info, edge->mutable_source());
+  edge->mutable_destination()->CopyFrom(node_instance_);
+
+  auto protocol = request_info.request_protocol;
+  if (protocol == "http" || protocol == "HTTP") {
+    edge->set_protocol(TrafficAssertion_Protocol_PROTOCOL_HTTP);
+  } else if (protocol == "https" || protocol == "HTTPS") {
+    edge->set_protocol(TrafficAssertion_Protocol_PROTOCOL_HTTPS);
+  } else if (protocol == "grpc" || protocol == "GRPC") {
+    edge->set_protocol(TrafficAssertion_Protocol_PROTOCOL_GRPC);
+  } else {
+    edge->set_protocol(TrafficAssertion_Protocol_PROTOCOL_TCP);
+  }
+
+  if (current_request_->traffic_assertions_size() >
+      max_assertions_per_request_) {
+    flush();
```

Comment:
Are we just dropping edges here without reporting? Should we log a warning at least?

Reply:
No. We shouldn't be dropping edges. I don't believe there is a hard upper-limit on the server side, and the limit here is just to keep the requests somewhat limited in size.